src/sqlfluff/core/config/fluffconfig.py

```diff
@@ -168,6 +169,17 @@ def __setstate__(self, state: Dict[str, Any]) -> None:  # pragma: no cover
         # which should also only be used in the main thread rather than child
         # processes.
 
+    def copy(self) -> FluffConfig:
+        """Returns a copy of the FluffConfig.
+
+        This creates a shallow copy of most of the config, but with a deep copy of the
+        `_configs` dictionary.
+        """
+        configs_attribute_copy = deepcopy(self._configs)
+        config_copy = copy(self)
+        config_copy._configs = configs_attribute_copy
```

Comment:
There's one failing test in the main suite, which I think will be solved by moving across a reference to the `templater_obj` like this:

```suggestion
        config_copy._configs = configs_attribute_copy
        # During the initial `.copy()`, we use the same `__reduce__()` method
        # which is used during pickling. The `templater_obj` doesn't pickle
        # well so is normally removed, but it's ok for us to just pass across
        # the original object here as we're in the same process.
        config_copy.templater_obj = self.templater_obj
```

Reply:
I believe there may be a problem with the `FluffConfig.__getstate__` method causing some breakage here as well. If that method is called, the original config will lose its `templater_obj`, which I assume is not intentional.
Example:
```python
config._configs["core"].get("tempalter_obj") # object exists
partial_config_copy = config.__getstate__()
config._configs["core"].get("templater_obj") # returns None
```
This occurs because of the `state["_configs"]["core"].pop("templater_obj", None)` line, which mutates the original config.

This is problematic when the config is held as an attribute within the linter and is used more than once (or if a user were to use the config in multiple parses even without maintaining a single `Linter` object.

Reply:
Hmmmm good point. 🤔 

I've got an idea. Hold up a sec...